src/Language/PureScript/TypeChecker/Entailment.hs

```diff
@@ -381,80 +390,93 @@ entails SolverOptions{..} constraint context hints =
         subclassDictionaryValue dict className index =
           App (Accessor (mkString (superclassName className index)) dict) valUndefined
 
-    solveCoercible :: Environment -> [SourceType] -> Maybe [TypeClassDict]
-    solveCoercible env [a, b] = do
+    solveCoercible :: Environment -> [SourceType] -> [SourceType] -> m (Maybe [TypeClassDict])
+    solveCoercible env kinds [a, b] = runMaybeT $ do
       let tySynMap = typeSynonyms env
           kindMap = types env
-          replaceTySyns = either (const Nothing) Just . replaceAllTypeSynonymsM tySynMap kindMap
-      a' <- replaceTySyns a
-      b' <- replaceTySyns b
+          replaceTySyns = replaceAllTypeSynonymsM tySynMap kindMap
+      (a', kind) <- lift $ replaceTySyns a >>= kindOf
+      (b', kind') <- lift $ replaceTySyns b >>= kindOf
+      lift $ unifyKinds kind kind'
```

Comment:
I'm sure this is a gap in my own understanding, but it feels a little bit strange to me that we aren't capturing any information as a result of unifying kinds here. Is it possible that we learn something about either `kind` or `kind'` that we didn't know until we reached this line? If so, should that information be carried forward?

Reply:
`unifyKinds` returns a `(MonadError MultipleErrors m, MonadState CheckState m) => m ()` so I’m not sure how to extract any information from a successful unification 🤔 

If unifying kinds extends the current substitution perhaps we should apply it to the kinds we inferred and then refer to `[kind, kind']` rather than `kinds` in the returned type class dictionary? Or perhaps we should rather apply it to the types rewritten by `kindOf` earlier? I’m completely out of my depth here 😅

Reply:
Yeah, I’m a bit out of my depth too. @natefaubion I’d be interested to hear your thoughts on this if you have a moment.

Reply:
`unifyKinds` does extend the substitution and at some point it must be applied, yes. I believe it's always applied at the top of the solver loop.

Reply:
Do you think that suggests that it might be worth exporting `apply` from TypeChecker.Kinds to use here?

Reply:
Ah, I suppose we should resolve the other discussion (https://github.com/purescript/purescript/pull/3893#discussion_r479778954) first.

Reply:
`applySubstitution` as is used everywhere else is fine. I think there's only one in the kind-checker for convenience or module dependencies or something.